### ISSUE TITLE
Can the stack memory usage limit be improved?

### ISSUE THREAD

**Issue by bitjson** [bitjson/bch-vm-limits#10]
Just opening this issue to solicit feedback from node/VM implementation developers: **can we improve the proposed stack memory usage limit**?

As [discussed on Bitcoin Cash Research](https://bitcoincashresearch.org/t/chip-2021-05-targeted-virtual-machine-limits/437/12?u=bitjson):

> My biggest uncertainty with the current proposal is the stack memory usage limit – it’s relatively expensive and inelegant to measure/enforce, so I’d especially love to see other interested developers research and review that part.
> 
> **I’d love to see node implementation developers experiment with how the VM can most simply and efficiently limit stack memory usage.**
> 
> The current proposal is essentially: `sum_all_stack_items + stack_depth <= 130,000` . That limit is rounded up from a max of `125,320 + 242` in what we believe is [the most memory-expensive script currently possible](https://github.com/bitjson/bch-vm-limits/#stack-memory-usage-limit).
> 
> The issue here is that we need to check memory usage after every operation in which memory usage might have increased.
> 
> Instead of the simple, incrementing counter that is currently used ( `++maxOperationsPerEvaluation < 201` ), we have to accumulate the sizes of all stack memory items, then add the length of the stack.
> 
> The calculation is not terribly expensive, and it can be optimized in many ways – e.g. many operations can skip the check, stack memory usage at >6 depth can be cached across operations, etc., but the current formulation still strikes me as less elegant than past improvements to the VM.
> 
> It’s possible that this is the best we can do, but I’d like to see some other people review the issue. **Some questions** :
> 
> * **Stack depth** : Do we actually need to limit stack depth? `MAX_SCRIPT_SIZE` is 10,000 bytes (and for standardness, `MAX_TX_IN_SCRIPT_SIG_SIZE` is only 1,650 bytes), so with `OP_3DUP` , we’re looking at a maximum depth of ~30k for transactions first heard in blocks (and only ~5k for transactions first heard over the network). Can we save cycles and complexity by just skipping that? (Are we ok with the protocol requiring stack + `OP_ROLL` implementations to be reasonably optimized?) And if not, do we need to multiply depth by some number >1 to represent the realistic overhead of a stack item?
> * **Memory usage** : Is there a more elegant way to limit stack memory usage? Rather than checking total usage after every operation in which it may have increased, is there a more efficient, less direct strategy?
> 
> If other developers with an interest in low-level optimization want to take a look at this, I’d love to hear your thoughts!

And [later](https://bitcoincashresearch.org/t/chip-2021-05-targeted-virtual-machine-limits/437/24?u=bitjson):

> Just wanted to link to recent related discussion on limiting memory usage: https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2023-October/022082.html
> 
> The patch bundles the limiting of total stack item memory usage into the stack depth limit.
> 
> Not sure if that ends up being simpler in most implementations than this CHIP’s current approach, but I wonder if that strategy opens up more opportunities for skipping the validation in common cases (so implementations can check memory usage less often).
> 
> Anyways, this question is still the main item that I personally think needs more research for this CHIP to be ready

**Comment by tomFlowee** [comment]
The discussion continued on bch-r. Maybe best to not fragment it?

https://bitcoincashresearch.org/t/chip-2021-05-targeted-virtual-machine-limits/437/32?u=tom

**Comment by A60AB5450353F40E** [comment]
This is related to https://github.com/bitjson/bch-vm-limits/issues/15

What's the point of 130kB limit if implementers will likely end up using 10MB per VM instance anyway, just so they don't have to juggle things in that tight space of 130kB.

If you have depth limit of 1000 and item limit of 10kB then it's simple to enforce, too: if result of any operation would exceed stack item size then fail the TX. If any operation would exceed the count of alt+main stacks, then fail the TX. No need to track total stack size if there's no need to enforce a limit on the total stack size. It'll still be bounded by max depth x max item size.

**Comment by tomFlowee** [comment]
As a full node dev myself, by far the simplest approach is to just increase the per-stack-item as the CHIP did.

and contrary to the current version, I think that is all that needs to be done. The 130kb limit that is then added is just creating more code for no reason.

The basic math is thus;

at any single point in time a single thread will only ever have 1 stack and one alt-stack. So the worst case is that a single thread pre-allocates 20MB of memory for script validation. On modern hardware this is peanuts. Even on a PI4 this is not an issue.

As an aside, I think that the pre-allocation is going to be something to do anyway if scripts become more stack-heavy in usage as the 'malloc' (which is today used for every stack-item) is extremely expensive compared to a pre-allocation.

Anyway, that is my point of view as an implementor.

**Comment by cculianu** [comment]
>  it’s relatively expensive and inelegant to measure/enforce, 
> Rather than checking total usage after every operation in which it may have increased, is there a more efficient, less direct strategy?

it's trivial and cheap to measure and enforce. Anybody claiming that it's expensive, complex, weird, etc... is just wrong, sorry. :P 

To me, 130kb seems reasonable esp. since it mimics previous limits in a different way.  

> (so implementations can check memory usage less often).

Again, this is super trivial. It's ridiculous anybody is "afraid" of this in any way.  It's.. just keeping an integer updated. Anything else the VM does (such as allocate, which it does quite often for every op, pretty much) is like 10,000x more expensive.

> Can we save cycles and complexity by just skipping that? (in reference to enforcing stack depth)

The number of cycles to be saved in not enforcing stack depth should not be a consideration. Enforcing stack depth is almost immeasurably cheap. Just an int comparison in C++.  Everything else the VM is doing (allocating, etc) is like 10000x more expensive.  It's a drop in the ocean.

---

FWIW, I would err on the side of caution and leave a 130kb memory limit in place in the spec, since it mimics current limits in a different way, but at the same time opens up new possibilities anyway, without being overly liberal.

Of course, the proof would be in the pudding with benchmarks and perf analysis and mem usage analysis in a running process, and whatnot...

For all we know the difference between *enforcing* 130KB and allowing *implicit* 10MB limit is tiny/non-existant.. or is huge and significant.  We would need to try and validate the most pathological set of txns or set of blocks imaginable to truly know what the impact would be in going from 130KB -> 10MB.

**Comment by cculianu** [comment]
> 'malloc' (which is today used for every stack-item) is extremely expensive compared to a pre-allocation.

Debateable. Note you would have to re-invent the wheel and do your own jerry-rigged and likely slower allocator anyway, since stack items are dynamically sized.  And if you decide to not write an allocator and always just use 10kb blobs for each stack item thinking you are clever, then you may end up being slower anyway than if you had used an allocator.. due to non-locality and thrashing your cache...  and more risk of kernel page faults due to sparser (mostly empty) page data in the common case.

Also note that BCHN uses jemalloc which has per-thread arenas anyway and so allocation largely doesn't take any locks.

Also allso note if you assume 20MB static stacks * 16 threads -> you are asking the OS for 320MB from the get-go. Now on most OS's it won't actually give you those pages anyway until you use them, but if stack items are sparse (that is, they are static 10KB blobs), you will be *eating more memory than had you used an allocator*, causing **more page faults**, and for each page fault you switch to the kernel, which has to do... *ta da*... **an allocation anyway** (it just happens in the kernel!).  

And.. That's slightly extra memory that the OS could have used for disk cache, other things, etc.  Just because of religious fear of userspace allocators.. :)

---

TL;DR: In my experience programs that think they are being clever and don't use allocators but instead used statically sized blobs end up eating more memory than they otherwise would have and rarely gain much in terms of performance.

**Comment by A60AB5450353F40E** [comment]
Nit: the new implicit limit would be 10MB (1k items x 10kB max. item size) because limit is for combined size: [`if (stack.size() + altstack.size() > MAX_STACK_SIZE) {`](https://gitlab.com/bitcoin-cash-node/bitcoin-cash-node/-/blob/master/src/script/interpreter.cpp#L1703)

Let's not forget that stack is not the only thing that uses memory, there's also the TX context: whole TX + prevouts. This can add up to something like 180 MB for a pathological case of someone with access to hashrate preparing a bunch of non-standard 10k long UTXOs and spending as much of them as can fit in a 1MB non-standard TX.

**Comment by tomFlowee** [comment]
> Anybody claiming that it's expensive, complex, weird, etc... is just wrong, sorry.

I respect you, and your obvious coding skills, but I think here we may have different limits on what is "complex".

Calin added a new class in the [PR](https://gitlab.com/bitcoin-cash-node/bitcoin-cash-node/-/merge_requests/1827) to BCHN for this CHIP. The class StackSentinel doesn't need to exist if the 130KB rule doesn't exist. So the complexity is actually quite easy to measure. Calin coded it. We start with a test of nearly 300 lines and the code itself is another 100 lines.

So, sure, the concept may be simple. Count something. So one integer, done. "keep an integer updated" But in practice it's never that simple. As the 400 lines added to BCHN show very clearly.

> Debateable. [long talk about how not to do memory alloc].

I agree with parts of it, but it makes no sense to debate on a not relevant issue. If you want I'll show some code to show how this is actually absurtly cheap to do and avoids all the pitfalls you are talk about. I guess just showing you the simple code on how we can make validation much faster is something useful at one point. But, again, largely irrelevant to this discussion.

**Comment by bitjson** [comment]
Hi all! Cleaning up issues now that we have two implementations of the latest proposal. 🚀

The answer to this issue was a resounding **yes: we can do far better than the initially-proposed, 130KB memory limit**. 

The latest approach limits "bytes pushed" in the course of evaluating contracts to a reasonable density based just above the existing maximum. This is both faster/simpler to enforce (rather than tracking or summing up stack size after each operation) and more comprehensive, since it prevents any kind of abuse that "moves bytes around at excessive rates", regardless of the high-water mark of memory usage during the evaluation (e.g. `OP_3DUP OP_3DUP OP_2DROP OP_2DROP OP_2DROP`). More detail in [Rationale: Limitation of Pushed Bytes](https://github.com/bitjson/bch-vm-limits/blob/b476674575e0f5d2ccdf60d5091a3b882b38b00c/rationale.md#limitation-of-pushed-bytes).

Thanks again to @cculianu, @tomFlowee, and @A60AB5450353F40E for helping to identify issues here with the initial approach. I'm going to mark this issue as resolved, but please comment if I should re-open.
